File: backend/ws-service/src/main/java/com/example/wsservice/kafka/producer/KafkaProducerService.java
package com.example.wsservice.kafka.producer;

import com.example.wsservice.chat.dto.ChatMessage;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class KafkaProducerService {

    private final KafkaTemplate<String, ChatMessage> kafkaTemplate;
    private static final String TOPIC = "topic-chat-save";
    private static final String READ_TOPIC = "topic-chat-read";

    public void sendMessage(String key, ChatMessage message) {
        kafkaTemplate.send(TOPIC, key, message)
                .whenComplete((result, ex) -> {
                    if (ex == null) {
                        log.info("Message sent successfully: {}", result.getRecordMetadata().topic() + " / " + message);
                    } else {
                        log.info("Message sent failed: {}", ex.getMessage());
                    }
                });
    }

    public void sendReadMessage(String key, ChatMessage message) {
        kafkaTemplate.send(READ_TOPIC, key, message)
                .whenComplete((result, ex) -> {
                    if (ex == null) {
                        log.info("Read sent successfully: {}", result.getRecordMetadata().topic() + " / " + message);
                    } else {
                        log.info("Read sent failed: {}", ex.getMessage());
                    }
                });
    }
}

File: backend/ws-service/src/main/kotlin/com/example/wsservice/kafka/producer/KafkaProducerService.kt
package com.example.wsservice.kafka.producer

import com.example.wsservice.chat.dto.ChatMessage
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service

@Service
class KafkaProducerService(private val kafkaTemplate: KafkaTemplate<String, ChatMessage>) {

    private val log = LoggerFactory.getLogger(KafkaProducerService::class.java)

    fun sendMessage(key: String, message: ChatMessage) = send(TOPIC, key, message, "Message")

    fun sendReactionMessage(key: String, message: ChatMessage) = send(REACTION_TOPIC, key, message, "Reaction")

    fun sendReadMessage(key: String, message: ChatMessage) = send(READ_TOPIC, key, message, "Read")

    private fun send(topic: String, key: String, message: ChatMessage, label: String) {
        kafkaTemplate.send(topic, key, message).whenComplete { result, ex ->
            if (ex == null) {
                log.info("{} sent successfully: {}", label, result.recordMetadata.topic() + " / " + message)
            } else {
                log.info("{} sent failed: {}", label, ex.message)
            }
        }
    }

    companion object {
        private const val TOPIC = "topic-chat-save"
        private const val READ_TOPIC = "topic-chat-read"
        private const val REACTION_TOPIC = "topic-chat-reaction"
    }
}
